### ISSUE TITLE
Wrong results when applying group_by().agg() to guvectorized function

### ISSUE THREAD

**Issue by soerenwolfers** [pola-rs/polars#14507]
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

I'm trying to compute a running sum where I clip at each step (so *not* cumsum().clip() and *not* clip().cumsum()) of the "values" column, per-"product".

I tried
```
import polars as pl
import numba

df = pl.DataFrame(
    {
        "product": ['a', 'a', 'a', 'a', 'b', 'b', 'b', 'b'],
        "values": [1.0, -1, 2, 3, -2, -2, 3, 0],
        "upper": [1.0, 2, 2, 2, 2, 2, 2, 1],
        "lower": [-1.0, -1, -1, -2, -2, -1, -1, -1],
    }
)

@numba.guvectorize([(numba.float64[:], numba.float64[:], numba.float64[:], numba.float64[:])], '(n),(n),(n)->(n)')
def bounded_cumulative_sum_2(arr: np.ndarray, upper: np.ndarray, lower: np.ndarray, out: np.ndarray):
    n = arr.shape[0]
    out[0] = arr[0]
    for i in range(1, n):
        out[i] = min(max(out[i - 1] + arr[i], lower[i]), upper[i])

df.group_by('product').agg(
    pl.all(),
    bounded_cumulative_sum_2(pl.col('values'), pl.col('upper'), pl.col('lower')).alias('bounded_cumulative_sum'),
).explode('bounded_cumulative_sum', 'values', 'upper', 'lower')
```

which prints
```
┌─────────┬────────┬───────┬───────┬────────────────────────┐
│ product ┆ values ┆ upper ┆ lower ┆ bounded_cumulative_sum │
│ ---     ┆ ---    ┆ ---   ┆ ---   ┆ ---                    │
│ str     ┆ f64    ┆ f64   ┆ f64   ┆ f64                    │
╞═════════╪════════╪═══════╪═══════╪════════════════════════╡
│ a       ┆ 1.0    ┆ 1.0   ┆ -1.0  ┆ 1.0                    │
│ a       ┆ -1.0   ┆ 2.0   ┆ -1.0  ┆ 0.0                    │
│ a       ┆ 2.0    ┆ 2.0   ┆ -1.0  ┆ 2.0                    │
│ a       ┆ 3.0    ┆ 2.0   ┆ -2.0  ┆ 2.0                    │
│ b       ┆ -2.0   ┆ 2.0   ┆ -2.0  ┆ 0.0                    │
│ b       ┆ -2.0   ┆ 2.0   ┆ -1.0  ┆ -1.0                   │
│ b       ┆ 3.0    ┆ 2.0   ┆ -1.0  ┆ 2.0                    │
│ b       ┆ 0.0    ┆ 1.0   ┆ -1.0  ┆ 1.0                    │
```

### Log output

```shell
keys/aggregates are not partitionable: running default HASH AGGREGATION
```

### Issue description

The output is wrong (at the start of the `b`-group the bounded_cumulative_sum should be -2, not 0) because it applies the function to the entire dataframe, not to the respective groups. 

PS: for my attempt to even work, I needed last weeks polars-0.20.8, but that's just because polars didn't accept guvectorize functions with more than two inputs before; my problem/bug is the same for functions with fewer inputs

### Expected behavior

```
  product  values  upper  lower  bounded_cumulative_sum
0       a     1.0    1.0   -1.0                     1.0
1       a    -1.0    2.0   -1.0                     0.0
2       a     2.0    2.0   -1.0                     2.0
3       a     3.0    2.0   -2.0                     2.0
4       b    -2.0    2.0   -2.0                    -2.0
5       b    -2.0    2.0   -1.0                    -1.0
6       b     3.0    2.0   -1.0                     2.0
7       b     0.0    1.0   -1.0                     1.0
```

### Installed versions

<details>

```
--------Version info---------
Polars:               0.20.8
Index type:           UInt32
Platform:             Linux-5.15.0-76-generic-x86_64-with-glibc2.35
Python:               3.10.6 (main, May 29 2023, 11:10:38) [GCC 11.3.0]

----Optional dependencies----
adbc_driver_manager:  <not installed>
cloudpickle:          3.0.0
connectorx:           <not installed>
deltalake:            <not installed>
fsspec:               <not installed>
gevent:               <not installed>
hvplot:               <not installed>
matplotlib:           3.7.2
numpy:                1.26.3
openpyxl:             <not installed>
pandas:               2.1.4
pyarrow:              14.0.2
pydantic:             <not installed>
pyiceberg:            <not installed>
pyxlsb:               <not installed>
sqlalchemy:           <not installed>
xlsx2csv:             <not installed>
xlsxwriter:           <not installed>
```

</details>

**Comment by itamarst** [comment]
A simplified reproducer:

```python
import numpy as np
import polars as pl
import numba

df = pl.DataFrame(
    {
        "product": ["a", "a", "a", "a", "b", "b", "b", "b"],
        "values": [1.0, -1, 2, 3, -2, -2, 3, 0],
    }
)


@numba.guvectorize(
    [(numba.float64[:], numba.float64[:])],
    "(n)->(n)",
)
def expects_grouped(arr: np.ndarray, out: np.ndarray):
    if arr.shape == (4,):
        # What we expect, given the group_by():
        for i in range(len(arr)):
            out[i] = arr[i] * 100
    else:
        # Store the actual size so we can see it:
        for i in range(len(arr)):
            out[i] = arr.shape[0]


result = df.group_by("product").agg(
    pl.all(),
    expects_grouped(pl.col("values")).alias("expects_grouped"),
)
print(result)
```

**Comment by itamarst** [comment]
So after poking around my guess is that what happens is that the `ufunc` is called _before_ the grouping is done, and then grouping happens post-facto. The other alternative is group first, call later separately on each batch.

If you have a stateless function (that only looks one item at a time), the order doesn't matter. For the given `ufunc` the batch matters, however, so doing things in this order gives unexpected results.

**Comment by soerenwolfers** [comment]
FWIW the numpy docs at https://numpy.org/doc/stable/reference/c-api/generalized-ufuncs.html# seem to be clear that the function in computing is in fact a legal generalized ufunc, but maybe polars implicitly assumes everything is a regular, non-generalized, ufunc?

**Comment by itamarst** [comment]
Using `map_elements()` does what @soerenwolfers expected, I think:

```python
result = df.group_by("product").agg(
    pl.all(),
    pl.col("values").map_elements(expects_grouped).alias("expects_grouped"),
)
```

**Comment by soerenwolfers** [comment]
Not clear to me how this would work with my initial example though, which requires three column inputs. Also, the initial report probably still constitutes a bug?

**Comment by itamarst** [comment]
@soerenwolfers this seems to work:

```python
print(
    df.group_by("product")
    .agg(
        pl.all(),
        pl.struct("values", "upper", "lower")
        .map_batches(
            lambda combined: bounded_cumulative_sum_2(
                combined.struct.field("values"),
                combined.struct.field("upper"),
                combined.struct.field("lower"),
            )
        )
        .alias("bounded_cumulative_sum"),
    )
    .explode("bounded_cumulative_sum", "values", "upper", "lower")
)
```

It gives:

```
shape: (8, 5)
┌─────────┬────────┬───────┬───────┬────────────────────────┐
│ product ┆ values ┆ upper ┆ lower ┆ bounded_cumulative_sum │
│ ---     ┆ ---    ┆ ---   ┆ ---   ┆ ---                    │
│ str     ┆ f64    ┆ f64   ┆ f64   ┆ f64                    │
╞═════════╪════════╪═══════╪═══════╪════════════════════════╡
│ a       ┆ 1.0    ┆ 1.0   ┆ -1.0  ┆ 1.0                    │
│ a       ┆ -1.0   ┆ 2.0   ┆ -1.0  ┆ 0.0                    │
│ a       ┆ 2.0    ┆ 2.0   ┆ -1.0  ┆ 2.0                    │
│ a       ┆ 3.0    ┆ 2.0   ┆ -2.0  ┆ 2.0                    │
│ b       ┆ -2.0   ┆ 2.0   ┆ -2.0  ┆ -2.0                   │
│ b       ┆ -2.0   ┆ 2.0   ┆ -1.0  ┆ -1.0                   │
│ b       ┆ 3.0    ┆ 2.0   ┆ -1.0  ┆ 2.0                    │
│ b       ┆ 0.0    ┆ 1.0   ┆ -1.0  ┆ 1.0                    │
└─────────┴────────┴───────┴───────┴────────────────────────┘
```

**Comment by deanm0000** [comment]
@itamarst Some background, it used to be that `map_batches` didn't respect groups at all. A few of us complained and ritchie changed the behavior and added an elementwise flag. The default is false so that groups will work. If you look at all the numpy and scipy ufuncs, they're all elementwise. In the `__array_ufunc__` implementation I made a check so that if the `ufunc` input is specifically `np.ufunc` then it would set the elementwise flag to true.

Since then, there seems to be a new version of either numba or numpy (I haven't figured out, and don't plan to look for, that specific change) where the `ufunc` parameter that `__array_ufunc__` receives looks exactly like a native numpy ufunc even though the function in our global environment has a class of `numba.np.ufunc.gufunc.GUFunc` and thusly _can be easily_ differentiated from a native numpy ufunc.

I made a [PR](https://github.com/pola-rs/polars/pull/15916) that changed that check to look for the ufunc.signature to be None.
